src/main/java/appeng/client/gui/implementations/GuiInterfaceTerminal.java

```diff
@@ -339,50 +304,376 @@ protected void actionPerformed(final GuiButton btn) {
+     *
+     * @param section   the section to render
+     * @param viewY     current y coordinate relative to gui
+     * @param relMouseX transformed mouse coords relative to viewport
+     * @param relMouseY transformed mouse coords relative to viewport
+     * @return the height of the section rendered in viewport coordinates, max of viewHeight.
+     */
+    private int drawSection(IfaceSection section, int viewY, int relMouseX, int relMouseY) {
+        int title;
+        int renderY = 0;
+        final int sectionBottom = viewY + section.getHeight() - 1;
+        final int fontColor = GuiColors.InterfaceTerminalInventory.getColor();
+        /*
+         * Render title
+         */
+        GL11.glTranslatef(0.0f, 0.0f, 50f);
+        bindTexture(BACKGROUND);
+        if (sectionBottom > 0 && sectionBottom < IfaceSection.TITLE_HEIGHT) {
+            /* Transition draw */
+            drawTexturedModalRect(
+                    0,
+                    0,
+                    VIEW_LEFT,
+                    HEADER_HEIGHT + IfaceSection.TITLE_HEIGHT - sectionBottom,
+                    VIEW_WIDTH,
+                    sectionBottom);
+            fontRendererObj.drawString(section.name, 2, sectionBottom - IfaceSection.TITLE_HEIGHT + 2, fontColor);
+            title = sectionBottom;
+        } else if (viewY < 0) {
+            /* Hidden title draw */
+            drawTexturedModalRect(0, 0, VIEW_LEFT, HEADER_HEIGHT, VIEW_WIDTH, IfaceSection.TITLE_HEIGHT);
+            fontRendererObj.drawString(section.name, 2, 2, fontColor);
+            title = 0;
+        } else {
+            /* Normal title draw */
+            drawTexturedModalRect(0, viewY, VIEW_LEFT, HEADER_HEIGHT, VIEW_WIDTH, IfaceSection.TITLE_HEIGHT);
+            fontRendererObj.drawString(section.name, 2, viewY + 2, fontColor);
+            title = 0;
+        }
+        GL11.glTranslatef(0.0f, 0.0f, -50f);
+        GL11.glColor4f(1.0f, 1.0f, 1.0f, 1.0f);
+
+        Iterator<IfaceEntry> visible = section.getVisible();
+        while (visible.hasNext()) {
+            if (viewY < viewHeight) {
+                renderY += drawEntry(
+                        visible.next(),
+                        viewY + IfaceSection.TITLE_HEIGHT + renderY,
+                        title,
+                        relMouseX,
+                        relMouseY);
+            } else {
+                IfaceEntry entry = visible.next();
+                entry.dispY = -9999;
+                entry.optionsButton.yPosition = -1;
+            }
+        }
+        return IfaceSection.TITLE_HEIGHT + renderY;
+    }
 
-        for (int x = 0; x < this.rows; x++) {
-            this.drawTexturedModalRect(offsetX, offsetY + 53 + x * 18, 0, 52, this.xSize, 18);
+    /**
+     * Draws the entry. In practice it just draws the slots + items.
+     *
+     * @param viewY the gui coordinate z
+     */
+    private int drawEntry(IfaceEntry entry, int viewY, int titleBottom, int relMouseX, int relMouseY) {
```

Comment:
Correct if I'm wrong but I think behaviour with green hilighter for founded patterns lost in new implementation